BIP 434: Peer Feature Negotiation (#2076)
bitcoin/bips: bip-peer-feature-negotiation.md

```
  BIP: ?
  Layer: Peer Services
  Title: Peer Feature Negotiation
  Author: Anthony Towns <aj@erisian.com.au>
  Comments-Summary: No comments yet.
  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-XXXX
  Status: Draft
  Type: Standards Track
  Created: 2025-12-06
  License: BSD-3-Clause
  Post-History: https://gnusha.org/pi/bitcoindev/aUUXLgEUCgGb122o@erisian.com.au/T/#u
                https://gnusha.org/pi/bitcoindev/20200821023647.7eat4goqqrtaqnna@erisian.com.au/
```

## Abstract

This BIP defines a peer-to-peer message that can be used for announcements
and negotiation related to support of new peer-to-peer features.

## Motivation

> Historically, new peer-to-peer protocol changes have been tied to

Reviewer: It's clear to us, but probably still good to define the acronym.

```suggestion
Historically, new peer-to-peer ("P2P") protocol changes have been tied to
```

bumping the protocol version, so that peers know to only attempt feature
negotiation with other peers that may have been upgraded to support the
feature. Coordinating the protocol version across implementations, when
different clients may have different priorities for features to implement,
is an unnecessary burden in the upgrade process for P2P features that do
not require universal support. And at a more philosophical level, having
the P2P protocol be [permissionlessly extensible][permless-extensible],
with no coordination required between implementations or developers
seems ideal for a decentralized system.

Many earlier P2P protocol upgrades were implemented as new messages
sent after a peer connection is setup (ie, after receipt of a `verack`
message by both sides). See [BIP 130 (sendheaders)][BIP130], [BIP 133

> (feefilter)][BIP133], [BIP 152 (compact blocks)][BIP152] for some

Reviewer: ```suggestion
(feefilter)][BIP133], and [BIP 152 (compact blocks)][BIP152] for
```

Nodes implementing this BIP:

 * MUST advertise a protocol version number `>= 70017`,
 * MUST NOT send `feature` messages to peers that advertise a protocol
   version number `< 70017`,
 * MUST accept `feature` messages received after the `version` message
   and before the `verack` message, and
 * MUST NOT send `feature` messages after sending the `verack` message.

In addition, nodes implementing this BIP:

 * SHOULD ignore unknown messages received after the `version` message
   and before the `verack` message,
 * MAY ignore `feature` messages sent after `verack`, and
 * MAY disconnect peers who send `feature` messages after `verack`.

Feature specifications based on this BIP:

 * MUST forbid sending messages it introduces after `verack` to a peer
   that has not indicated support for the feature via a `feature`
   message.

### `feature` message

The payload of the `feature` message contains exactly the following data:

| Type        | Name          | Description |
| ----------- | ------------- | ----------- |
| string      | `featureid`   | Unique identifier for the feature |
| byte-vector | `featuredata` | Feature-specific configuration data |

> The `featureid` is encoded in the usual way, that is as a `CompactSize`

Reviewer: ```suggestion
The `featureid` is encoded in the usual way, that is, as a `CompactSize`
```

specifying the string length, followed by that many bytes. The string
length MUST be between 4 and 80, inclusive. The string SHOULD include
only printable ASCII characters (ie, each byte should have a value between 32 and
126, inclusive).

Likewise, `featuredata` is encoded as a `CompactSize` specifying the
byte-vector size, followed by that many bytes. How these bytes are
interpreted is part of the feature's specification. The byte-vector size
MUST NOT be more than 512 bytes. Note that the `featuredata` field is not
optional, so if no data is required, an empty vector should be provided,
ie serialized as a `CompactSize` of 0.

Nodes implementing this BIP MAY disconnect peers that send `feature`
messages where the `feature` message's payload cannot be correctly
parsed (including having missing or additional data), even if they do

> not recognise the `featureid`.

Reviewer: It may be clearer to place this paragraph after the next one, as the "even if" phrase is confusing before reading the next paragraph.

    as soon as `feature BIP339v2` is received.
 6. Conversely, if the initiating peer sends `BIP339v3`, `BIP339v2`,
    `BIP339v1`, and the listening peer sends `BIP339v1`, `BIP339v2`,
    then the listening peer should select `BIP339v1` when `verack`
    is received, and the initiating peer should select `BIP339v1`
    as soon as `feature BIP339v1` is received.
 7. In most cases, implementations should simply advertise incompatible
    features in order from most recent to oldest, on the basis that
    the only reason to make incompatible updates is because there are
    significant improvements. Exceptions to that may occur when two
    incompatible features are both receiving active development, or
    when an implementation has only partially implemented the latest
    spec, and the older spec is better supported (and thus should be
    listed first, as the preferred protocol to adopt).

This approach may be appropriate when making substantial changes to a
deployed protocol and backwards compatibility is desirable on a short-term
basis, or when there is disagreement amongst implementations or users
as to which approach is most desirable.

> ## Considerations

Reviewer 2: Nit: I’m interpreting this as the stand-in for the Rationale section.

Author: Some of the rationale is in the motivation section ("what inspired the design"), some is in the spec itself (particularly the "feature negotiation" parts); the "considerations" section focuses on "why these design decisions vs other similarly plausible ones". [49](https://github.com/bitcoin/bips/blob/master/bip-0049.mediawiki#user-content-Considerations) (and 84), [90](https://github.com/bitcoin/bips/blob/master/bip-0090.mediawiki#user-content-Considerations), and [348](https://github.com/bitcoin/bips/blob/master/bip-0348.md#design-considerations) also use a Considerations section fwiw. I'm mostly following the structure of [Suhas's old draft](https://github.com/sdaftuar/bips/blob/2020-08-generalized-feature-negotiation/bip-p2p-feature-negotiation.mediawiki). I didn't rename it to Rationale because it didn't seem sensible to move all the rationale into that section, or to do the footnotes thing like [340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#user-content-Footnotes), and BIP3 specifies Rationale as optional, and that it's fine as long as you "consider and address as appropriate".

The advantage this approach has over defining new messages for each
feature is that the `featureid` can be much longer (at up to 80
bytes) than a message type id (which are limited to 12 bytes). With a
[BIP 324][BIP324] one-byte `message_type`, the overhead compared to that
approach is also kept small.

This approach is largely equivalent to adding a [payload to the `verack`
message][verack-payload] (eg, a vector of `featureid`, `featuredata`
pairs). It was chosen because:

 * it retains compatibility with any implementations that expect `verack`
   to have no payload;
 * it allows peers to process each feature request individually, rather than
   having to first load the configuration information for all features into
   memory at once (in order to validate the message's checksum), and then
   deal with each feature's configuration;
 * limiting the maximum message payload size you accept (eg to 4MB)
   does not limit the number of features you can accept; and
 * we have experience with negotiating features with individual messages,
   but no experience with doing so via `verack` payload.

A mild disadvantage compared to using a `verack` payload is that this

> approach allows the possibility of interactive feature negotiation prior

Reviewer: I may be confused.
```suggestion
approach does not allow the possibility of interactive feature negotiation prior
```

to `verack`. However interactive feature negotiation is always possible
simply by having the initiating peer disconnect and reconnect after
discovering the listening peer's supported features.

This specification attempts to maximise compatibility with implementations
that prefer to fully validate each message received:

 * `feature` messages, even for unknown features, must always be fully
   parseable into a `featureid` and `featuredata`
 * Ignoring unknown messages prior to `verack` is only a recommendation

>    nor a requirement, so compliant implementations may disconnect on an

Reviewer: "not" meant here?
```suggestion
   , not a requirement, so compliant implementations may disconnect on an
```
